deps/v8/src/api/api-wasm.cc

```diff
@@ -0,0 +1,317 @@
+// Copyright 2019 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "src/api/api-wasm.h"
+#include "src/base/memory.h"
+#include "src/common/globals.h"
+#include "src/api/api-inl.h"
+#include "src/objects/contexts.h"
+#include "src/objects/lookup.h"
+#include "src/execution/isolate.h"
+
+#define IGNORE(X) ((void) (X))
+
+namespace b = v8::base;
+namespace i = v8::internal;
+
+namespace v8 {
+namespace wasm {
+
+Memory::Memory(size_t pages, uint8_t* data) : pages_(pages), data_(data) {}
+Memory::Memory(const Memory& memory)
+    : pages_(memory.pages_), data_(memory.data_) {}
+size_t Memory::size() { return pages_ * PAGE_SIZE; }
+size_t Memory::pages() { return pages_; }
+uint8_t* Memory::data() { return data_; }
+
+Table::Table(void* tableObject) {
+  tableObject_ = (void*) new i::WasmTableObject(
```

Comment:
IIUC, this doing a copy construction of a new `i::WasmTableObject`. Was that the intention, or was the intention to keep a reference to the `i::WasmTableObject` passed in?

Either way, it would be more readable to extract the first cast to an assignment of its own:
```C++
auto otherTableObject = (i::WasmTableObject*) tableObject;
tableObject_ = new i::WasmTableObject(*otherTableObject);
```

The cast to `void*` should be unnecessary.

Reply:
Quick explanation for the reason this is `void*` here: since I need to define this type in the V8 headers, I can't depend on anything in the `v8::internal` namespace. Thus I store it as a `void*` and cast it when necessary. I'm a little unsure of whether the copy here is correct, but IIRC the passed in value is just a reference to a stack-allocated table object that needs to be copied so that we can store it for later. The complication here is that the `Context` object may get used more than once or stored by the callee, so we have to be careful just creating a `Context` and then deleting it. Would like to discuss this in depth a little bit and find a better solution